src/osd/PrimaryLogPG.cc

```diff
@@ -8501,16 +8501,26 @@ void PrimaryLogPG::_do_rollback_to(OpContext *ctx, ObjectContextRef rollback_to,
   t->clone(soid, rollback_to_sobject);
   t->add_obc(rollback_to);
 
+  if (obs.oi.size > 0) {
+    // change most recent clone_overlap to [0, obs.oi.size] before rollback to snap
+    if (ctx->new_snapset.clones.size() > 0) {
+      interval_set<uint64_t> rollback_overlaps;
+      rollback_overlaps.insert(0, obs.oi.size);
+      interval_set<uint64_t> &newest_overlap =
+        ctx->new_snapset.clone_overlap.rbegin()->second;
+      newest_overlap.swap(rollback_overlaps);
```

Comment:
Modifying clone_overlaps directly here isn't appropriate -- that's what ctx->modified_ranges is for.  See the end of PrimaryLogPG::make_writeable.

Also, this logic looks wrong -- it looks like this last clone_overlap would indicate that the most recent clone and head *share* everything in the range [0, obs.oi-size).  Is that actually the intent?

In a larger sense, I'm not actually sure what about the logic below is wrong.  At a minimum, this PR needs a commit adding a unit test which fails with the current implementation.  It looks like you have a reproducer?